File: app/src/main/java/com/autohub/skln/activities/user/StudentHobbySelect.java
package com.autohub.skln.activities.user;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.autohub.skln.BaseActivity;
import com.autohub.skln.R;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.SetOptions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

import static com.autohub.skln.utills.AppConstants.HOBBY_DANCE;
import static com.autohub.skln.utills.AppConstants.HOBBY_DRUM;
import static com.autohub.skln.utills.AppConstants.HOBBY_GUITAR;
import static com.autohub.skln.utills.AppConstants.HOBBY_KEYBOARD;
import static com.autohub.skln.utills.AppConstants.HOBBY_MARTIAL;
import static com.autohub.skln.utills.AppConstants.HOBBY_PAINT;
import static com.autohub.skln.utills.AppConstants.KEY_PHONE_NUMBER;
import static com.autohub.skln.utills.AppConstants.KEY_STDT_HOBBIES;

public class StudentHobbySelect extends BaseActivity {

    @BindView(R.id.ivGuitar)
    ImageView ivGUitar;

    @BindView(R.id.ivpainting)
    ImageView ivPainting;

    @BindView(R.id.ivMartialArts)
    ImageView ivMartialArts;

    @BindView(R.id.ivDrum)
    ImageView ivDrum;

    @BindView(R.id.ivKeyboard)
    ImageView ivKeyboard;

    @BindView(R.id.ivDance)
    ImageView ivDance;

    private ArrayList<String> selectedHobbies;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_student_hobby_select);
        ButterKnife.bind(this);

        ivGUitar.setEnabled(false);
        ivPainting.setEnabled(false);
        ivMartialArts.setEnabled(false);
        ivDance.setEnabled(false);
        ivDrum.setEnabled(false);
        ivKeyboard.setEnabled(false);

        selectedHobbies = new ArrayList<>();
    }

    @OnClick(R.id.btnNext)
    public void onNextClick() {
        StringBuilder stringBuilder = new StringBuilder();
        if (selectedHobbies.size() > 0) {
            stringBuilder.append(selectedHobbies.get(0));
            for (int i = 1; i < selectedHobbies.size(); i++) {
                stringBuilder.append(", ").append(selectedHobbies.get(i));
            }
        }

        if (stringBuilder.length() == 0) {
            showSnackError(R.string.choose_hobbies);
            return;
        }

        showLoading();

        Map<String, Object> user = new HashMap<>();
        user.put(KEY_STDT_HOBBIES, stringBuilder.toString());

        final Map<String, Object> user1 = new HashMap<>();
        user1.put(KEY_PHONE_NUMBER, getAppPreferenceHelper().getUserPhone());

        getFirebaseStore().collection(getString(R.string.db_root_students)).document(getFirebaseAuth().getCurrentUser().getUid()).set(user, SetOptions.merge())
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {

                        getFirebaseStore().collection(getString(R.string.db_root_all_users)).document(getFirebaseAuth().getCurrentUser().getUid()).set(user1, SetOptions.merge())
                                .addOnSuccessListener(new OnSuccessListener<Void>() {
                                    @Override
                                    public void onSuccess(Void aVoid) {
                                        hideLoading();
                                        getAppPreferenceHelper().setSignupComplete(true);
                                        Intent i = new Intent(StudentHobbySelect.this, StudentHomeActivity.class);
                                        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                                        startActivity(i);
                                        finish();
                                    }
                                })
                                .addOnFailureListener(new OnFailureListener() {
                                    @Override
                                    public void onFailure(@NonNull Exception e) {
                                        hideLoading();
                                        showSnackError(e.getMessage());
                                    }
                                });
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        hideLoading();
                        showSnackError(e.getMessage());
                    }
                });
    }

    @OnClick(R.id.LLGuitar)
    public void onGuitarClick() {
        if (!ivGUitar.isEnabled()) {
            selectedHobbies.add(HOBBY_GUITAR);
            ivGUitar.setEnabled(true);
        } else {
            selectedHobbies.remove(HOBBY_GUITAR);
            ivGUitar.setEnabled(false);
        }
    }

    @OnClick(R.id.LLPainting)
    public void onPaintingClick() {
        if (!ivPainting.isEnabled()) {
            selectedHobbies.add(HOBBY_PAINT);
            ivPainting.setEnabled(true);
        } else {
            selectedHobbies.remove(HOBBY_PAINT);
            ivPainting.setEnabled(false);
        }
    }

    @OnClick(R.id.LLMartialArts)
    public void onMartialArtClick() {
        if (!ivMartialArts.isEnabled()) {
            selectedHobbies.add(HOBBY_MARTIAL);
            ivMartialArts.setEnabled(true);
        } else {
            selectedHobbies.remove(HOBBY_MARTIAL);
            ivMartialArts.setEnabled(false);
        }
    }

    @OnClick(R.id.LLDrum)
    public void onDrumClick() {
        if (!ivDrum.isEnabled()) {
            selectedHobbies.add(HOBBY_DRUM);
            ivDrum.setEnabled(true);
        } else {
            selectedHobbies.remove(HOBBY_DRUM);
            ivDrum.setEnabled(false);
        }
    }

    @OnClick(R.id.LLkeyboard)
    public void onKeyBoardClick() {
        if (!ivKeyboard.isEnabled()) {
            selectedHobbies.add(HOBBY_KEYBOARD);
            ivKeyboard.setEnabled(true);
        } else {
            selectedHobbies.remove(HOBBY_KEYBOARD);
            ivKeyboard.setEnabled(false);
        }
    }

    @OnClick(R.id.LLDance)
    public void onDanceClick() {
        if (!ivDance.isEnabled()) {
            selectedHobbies.add(HOBBY_DANCE);
            ivDance.setEnabled(true);
        } else {
            selectedHobbies.remove(HOBBY_DANCE);
            ivDance.setEnabled(false);
        }
    }

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }

}

File: app/src/main/java/com/autohub/skln/activities/user/StudentHobbySelect.kt
package com.autohub.skln.activities.user

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import com.autohub.skln.BaseActivity
import com.autohub.skln.R
import com.autohub.skln.activities.user.fragments.HobbiesFragment
import com.autohub.skln.databinding.ActivityStudentHobbySelectBinding
import com.autohub.skln.listeners.ClassSelectionListner
import com.autohub.skln.models.HobbiesData
import com.autohub.skln.utills.AppConstants.*
import com.google.firebase.firestore.SetOptions
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper
import java.util.*

class StudentHobbySelect : BaseActivity(), ClassSelectionListner {
    override fun selectedClass(position: Int, isSecondSelected: Boolean, selectedhobby: String) {
        for (i in fragmentsList.indices) {
            if (position == i) {
                if (isSecondSelected) {
                    if (selectedHobbies.contains(selectedhobby)) {
                        selectedHobbies.remove(selectedhobby)
                        fragmentsList[i].updateFragment(isSecond = true, selected = false)

                    } else {
                        selectedHobbies.add(selectedhobby)

                        fragmentsList[i].updateFragment(isSecond = true, selected = true)

                    }

                } else {
                    if (selectedHobbies.contains(selectedhobby)) {
                        selectedHobbies.remove(selectedhobby)
                        fragmentsList[i].updateFragment(isSecond = false, selected = false)

                    } else {
                        selectedHobbies.add(selectedhobby)

                        fragmentsList[i].updateFragment(isSecond = false, selected = true)

                    }
                }


            }
        }
    }


    private var selectedHobbies: ArrayList<String> = ArrayList()
    private var mBinding: ActivityStudentHobbySelectBinding? = null
    private var hobbiesDataList: ArrayList<HobbiesData> = ArrayList()
    private var fragmentsList: ArrayList<HobbiesFragment> = ArrayList()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_student_hobby_select)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_student_hobby_select)
        mBinding!!.callback = this
        insertHobbiesData()

        var countList: ArrayList<String> = ArrayList()
        countList.add("1")
        countList.add("2")
        countList.add("3")

        getFragments(countList)
        var pagerAdapter = PagerAdapter(supportFragmentManager, fragmentsList)
        mBinding!!.viewpager.adapter = pagerAdapter
        mBinding!!.viewpager.offscreenPageLimit = 2
        mBinding!!.wormDotsIndicator.setViewPager(mBinding!!.viewpager)


    }

    private fun getFragments(countList: ArrayList<String>) {


        for (position in countList.indices) {

            if (position == 0) {
                fragmentsList.add(HobbiesFragment.newInstance(position, ArrayList<HobbiesData>(hobbiesDataList.subList(0, 2))))

            } else if (position == 1)
                fragmentsList.add(HobbiesFragment.newInstance(position, ArrayList<HobbiesData>(hobbiesDataList.subList(2, 4))))
            else
                fragmentsList.add(HobbiesFragment.newInstance(position, ArrayList<HobbiesData>(hobbiesDataList.subList(4, 6))))
        }
    }

    private fun insertHobbiesData() {
        hobbiesDataList.add(HobbiesData(R.color.guitar, HOBBY_GUITAR, R.drawable.guitar));
        hobbiesDataList.add(HobbiesData(R.color.keyboard, HOBBY_KEYBOARD, R.drawable.piano));
        hobbiesDataList.add(HobbiesData(R.color.materialarts, HOBBY_MARTIAL, R.drawable.attack));

        hobbiesDataList.add(HobbiesData(R.color.dance, HOBBY_DANCE, R.drawable.dancing))
        hobbiesDataList.add(HobbiesData(R.color.painting, HOBBY_PAINT, R.drawable.brush));

        hobbiesDataList.add(HobbiesData(R.color.drum, HOBBY_DRUM, R.drawable.drum));

    }

    fun onNextClick() {
        val stringBuilder = StringBuilder()
        if (selectedHobbies!!.size > 0) {
            stringBuilder.append(selectedHobbies!![0])
            for (i in 1 until selectedHobbies!!.size) {
                stringBuilder.append(", ").append(selectedHobbies!![i])
            }
        }

        if (stringBuilder.length == 0) {
            showSnackError(R.string.choose_hobbies)
            return
        }

        showLoading()

        val user = HashMap<String, Any>()
        user[KEY_STDT_HOBBIES] = stringBuilder.toString()

        val user1 = HashMap<String, Any>()
        user1[KEY_PHONE_NUMBER] = appPreferenceHelper.userPhone

        firebaseStore.collection(getString(R.string.db_root_students)).document(firebaseAuth.currentUser!!.uid).set(user, SetOptions.merge())
                .addOnSuccessListener {
                    firebaseStore.collection(getString(R.string.db_root_all_users)).document(firebaseAuth.currentUser!!.uid).set(user1, SetOptions.merge())
                            .addOnSuccessListener {
                                hideLoading()
                                appPreferenceHelper.setSignupComplete(true)
                                val i = Intent(this@StudentHobbySelect, StudentHomeActivity::class.java)
                                i.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                                startActivity(i)
                                finish()
                            }
                            .addOnFailureListener { e ->
                                hideLoading()
                                showSnackError(e.message)
                            }
                }
                .addOnFailureListener { e ->
                    hideLoading()
                    showSnackError(e.message)
                }
    }

    inner class PagerAdapter(fragmentManager: FragmentManager, fragmentsList: ArrayList<HobbiesFragment>) :
            FragmentStatePagerAdapter(fragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

        // 2
        override fun getItem(position: Int): Fragment {
            return fragmentsList[position]


        }

        override fun getPageWidth(position: Int): Float {
            return 0.5f
        }

        // 3
        override fun getCount(): Int {
            return fragmentsList.size
        }
    }


    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase))
    }

}
